### ISSUE TITLE
v20.9.3 hogs system resources

### ISSUE THREAD

**Issue by shinenelson** [sindresorhus/refined-github#3527]
### Description

The [latest release 20.9.3](https://github.com/sindresorhus/refined-github/releases/tag/20.9.3) that rolled into my Firefox Developer Edition ( v81.0b5 ) on an Ubuntu 20.04.1 LTS machine earlier today started hogging all of my system resources whenever a github page was loaded.

### Steps to Reproduce

Optional : create a [fresh profile on Firefox](https://support.mozilla.org/en-US/kb/profile-manager-create-remove-switch-firefox-profiles#w_creating-a-profile)
1. Install [Refined Github](https://addons.mozilla.org/en-US/firefox/addon/refined-github-/) extension from addons.mozilla.org
2. Open a Github page

### Expected Behavior

The page loads without issues.

### Actual Behaviour

1. Firefox warns that the extension is slowing the page down.
![Screenshot from 2020-09-04 05-58-13](https://user-images.githubusercontent.com/4771718/92186691-34a47980-ee75-11ea-8807-a21a8092dfc0.png)
2. CPU hits 100% ( on all cores ) and the fans start spinning very loudly
3. RAM starts to fill up until none is left.

Clicking the 'Stop it' button in the browser warning does not take immediate effect because the whole browser process is in some kind of a hung state. The only 2 ways ( that I could figure was ) to stop it is to either disable the add-on from the Add-ons Manager ( if the tab is already open; new tabs don't load the pages within the DOM ) or to kill the whole browser process.

### Test Environment

Operating System : Ubuntu 20.04.1 LTS
Browser Version : Firefox 80.0, Firefox Developer Edition ( 81.0b5 )
Extension Version : 20.9.3

**Comment by fregante** [comment]
Thanks for the detailed report, but does this happen on any page? If not, please leave a link where it happens. 

There have been some changes lately that might cause this, but we had just hoped they wouldn't — #3084

**Comment by fregante** [comment]
Also, can you try finding which version started causing this? https://addons.mozilla.org/en-US/firefox/addon/refined-github-/versions/

**Comment by fregante** [comment]
Also, more helpfully, can you try clicking on "Debug script" to see where it stops? This might take a minute to stop, but it will provide information about what's causing it.

**Comment by njibhu** [comment]
Same issue here on ArchLinux, FF 80.0.1
Actually as soon as I interact with the yellow bar, either to disable the addon on the page, stop the script or debug the whole Firefox UI stops responding and only solution is to kill Firefox.
I'll have to disable the extension for the time being.
I think the issue might have caused by the 20.9.3 update. Before this update everything was working perfectly on my side.

If there's any other way to give helpful information I'd be happy to provide them.

**Comment by swt30** [comment]
Same here on OSX Catalina, Firefox 81.0b4, Refined Github 20.9.3, on any github page. I'll hit debug and see what happens.

**Comment by noaione** [comment]
Same problem, it only occured on repository page.
I'm using version 20.9.3 on Chrome 85.0.4183.83

Looking at the dev console, the problem is because this line of code https://github.com/sindresorhus/refined-github/blob/master/source/features/cleanup-repo-filelist-actions.tsx#L38

```js
refined-github.js:9621 Uncaught TypeError: Cannot read property 'nextSibling' of null
```

Line 9621 on the packaged(?) script
```js
        Object(select_dom["a"])(".octicon-download", downloadButton).nextSibling.remove();
```

**Comment by shinenelson** [comment]
That would probably explain why this 'hang' happens only on the "Code" tab and doesn't happen on "Issues" or "Pull requests" tabs.

I was looking at the [changes that went into this release](https://github.com/sindresorhus/refined-github/compare/20.8.30...20.9.3) - it wasn't much but I found #3517 and somehow felt that the `observe` block and the `onetime(init)` could be the source of the problem; though I couldn't connect it to the `selector-observer` and https://bugzil.la/1422522

**Comment by yakov116** [comment]
I just tried to reproduce I did a fresh install (brand new computer) and I am not seeing the issue.

**Comment by yakov116** [comment]
@noaione I dont see the error. Do you have some kind of feature preview enabled? Or exact steps you are doing?

**Comment by shinenelson** [comment]
@yakov116 in my testing, I created a new Firefox profile to simulate a fresh install. The only extension I installed on the profile was the latest release of refined github. In fact, that was the only thing I did on that profile - my Steps to Reproduce in the description of the bug report describes it.

I opened the new profile, installed the Refined Github extension, and the install opens the Refined Github updates issue with no problem. Then I click on the Code tab on the page and the page almost loads completely before the warning is shown by the browser that the extension is slowing the page down. Then it goes on a resource hog.

**Comment by shinenelson** [comment]
@yakov116 in my testing, I created a new Firefox profile to simulate a fresh install. The only extension I installed on the profile was the latest release of refined github. In fact, that was the only thing I did on that profile - my Steps to Reproduce in the description of the bug report describes it.

I opened the new profile, installed the Refined Github extension, and the install opens the Refined Github updates issue with no problem. Then I click on the Code tab on the page and the page almost loads completely before the warning is shown by the browser that the extension is slowing the page down. Then it goes on a resource hog.

**Comment by yakov116** [comment]
@shinenelson can you try to back version to see if that helps?

**Comment by shinenelson** [comment]
like user njibhu commented, this started happening after the latest update. I had the extension enabled until then and it was working fine.

and like fregante commented, this was caused due to the change in #3517 with `selector-observer` which went into this latest release. Though the Firefox bug report that I had linked to in a previous comment ( which was sourceed from 3084 ) was **RESOLVED FIXED** 2 years ago, the issue could also come back again now.

**Comment by yakov116** [comment]
I understand and would like to attempt to debug it. Are you able to load a previous version of refined-github to see if that fixes the issue

**Comment by shinenelson** [comment]
I thought it was obvious from mine as well as njibhu's comments that previous versions would work fine until this latest release. I will still do a test for you on a fresh profile with the last version.

**Comment by yakov116** [comment]
Can you see if it freezes when you go to https://github.com/sindresorhus/refined-github/branches? Select over server is used there too (and many other pages implemented about a month ago)

**Comment by fregante** [comment]
Does disabling `cleanup-repo-filelist-actions` fix it?

**Comment by shinenelson** [comment]
>  Are you able to load a previous version of refined-github to see if that fixes the issue

Yes, it does. I rolled back to extension version 20.8.30 ( of course on a fresh new profile ) and could not reproduce the issue. This is not really a fix or a solution. It is only validating that the latest version is the offending version.

> Can you see if it freezes when you go to https://github.com/sindresorhus/refined-github/branches?

No, it does not. As (re-)confirmed with the test with the previous version, the offending feature is in this latest release.

> Does disabling `cleanup-repo-filelist-actions` fix it?

Yes, it does!

thinks to self : " _why did you have to disable the whole extension for one offending feature? why didn't you think of disabling the offending feature yourself?_ " :facepalm:

---

disabling the `cleanup-repo-filelist-actions` feature is a good workaround to this current problem rather than having to disable the whole extension. I don't want to miss out on the many features that the extension already provides because of one feature that kind of went rogue.

**Comment by fregante** [comment]
I'm publishing 20.9.4 for the time being, reverting https://github.com/sindresorhus/refined-github/pull/3517

**Comment by fregante** [comment]
Can you provide exact URLs where this happen? Does this happen on https://github.com/sindresorhus/refined-github?

**Comment by autra** [comment]
> Are you guys on Enterprise?

nope.

btw, I feel like there is an infinite loop somewhere. I'm pretty sure it's not connected to the fact observers' performance are not *super*.

It happens on https://github.com/tituspijean/docs.contribute for me (and other pages too).

**Comment by shinenelson** [comment]
noaione's https://github.com/sindresorhus/refined-github/issues/3527#issuecomment-687079196 indicate that it should be on every page that has the `.octicon-download` class which would mean `<repo-url>/tree/*` where the download button that uses https://download-directory.github.io is present.

> Are you guys on Enterprise?

not me.

**Comment by fregante** [comment]
> the download button that uses [download-directory.github.io](https://download-directory.github.io) is present

That button isn't affected, the selector is very specifically `get-repo details` which **only appears on the root** (`get-repo` being the custom dropdown element with the "clone/download" links)

**Comment by autra** [comment]
Aaaaannd I can't reproduce any more. I'm pretty sure it's an interaction with another addon, possibly octotree, because disabling and enabling it made the problem go away for me.

**Comment by shinenelson** [comment]
@autra : can you check whether your add-on got updated to v20.9.4? it was published a little earlier reverting #3517.

**Comment by kidonng** [comment]
@autra I have Octotree as well and cannot reproduce (on Chrome 86.0.4240.22)

**Comment by shinenelson** [comment]
> That button isn't affected, the selector is very specifically get-repo details which only appears on the root (get-repo being the custom dropdown element with the "clone/download" links)

@fregante is right. that was the change in #3517. my bad :pray:

so, all the root repository pages then? I think I had it reproduce on [josh/selector-observer](https://github.com/josh/selector-observer) too during my initial hangs and tests ( before I went hard-core debugging the issue )

**Comment by fregante** [comment]
The error in https://github.com/sindresorhus/refined-github/issues/3527#issuecomment-687079196 suggests that `document.querySelector('get-repo details .octicon-download')` doesn't exist. That's really strange because we don't remove it.

Can anyone post a screenshot of the whole viewport when it hangs? Perhaps GitHub is doing some A/B

**Comment by fregante** [comment]
I made it crash with:

1. Remove download icon
2. Rename `details` tag to `whatever`
3. Rename it back to `details`, it immediately hangs

**Comment by shinenelson** [comment]
I don't know if this will be of much help

![image](https://user-images.githubusercontent.com/4771718/92278535-0d9c8500-ef13-11ea-9a6b-78b4de2f0a22.png)

The `.octicon-download` button is there though.

**Comment by kidonng** [comment]
This is really strange, I can reproduce it when I didn't login into GitHub, but cannot when I logged in (both without setting token)

**Comment by fregante** [comment]
So indeed this loops inside the `add` function, probably because of the error. This is kinda of a new problem with `selector-observer` because if that's what happens every time the DOM changes we're screwed.

**Comment by kidonng** [comment]
@fregante Maybe just

```diff
- select('.octicon-download', downloadButton)!.nextSibling!.remove();
+ select('.octicon-download', downloadButton)?.nextSibling!.remove();
```

**Comment by fregante** [comment]
That just hides the error, we don't do that. We can however change the original selector to point directly to the SVG, which would fix this specific issue for the time being.

---

My guess is that we add the classes, triggering a new MutationEvent, but `selector-observer` doesn't get to deduplicate it because `add` throws an error.

**Comment by kidonng** [comment]
> My guess is that we add the classes, triggering a new MutationEvent, but `selector-observer` doesn't get to deduplicate it because `add` throws an error.

I tried to only remove the last line (`select('.octicon-download', downloadButton)!.nextSibling!.remove()`) and it didn't stuck

**Comment by autra** [comment]
> @autra : can you check whether your add-on got updated to v20.9.4? it was published a little earlier reverting #3517.

I'm still on v20.9.3

**Comment by fregante** [comment]
@kidonng probably just observe the download button directly:

```js
observe('get-repo details .octicon-download', {............})
```
